feat(router): allow passing an ActivatedRouteSnapshot for relative link resolution in Router.createUrlTree

daiscog opened the pull request:
The relativeTo property of the navigationExtras parameter in the createUrlTree method of the Router class may now be an ActivatedRoute or an ActivatedRouteSnapshot.
The latter is useful in CanActivate route guards when building an alternative UrlTree to navigate to, relative to the URL that the current navigation is attempting to access.

Fixes #22763

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?

As reported in Issue Number: #22763, when implementing a CanActivate route guard, there is no way to return a UrlTree relative to the route that would otherwise have been activated, as the current ActivatedRoute is the previous navigation, and the next navigation is only available via the ActivatedRouteSnapshot given to the guard.


## What is the new behavior?

This change allows passing an ActivatedRouteSnapshot to as the relativeTo value, thus enabling the following pattern in a route guard:

```ts
export class ExampleGuard implements CanActivate {
  constructor(private readonly router: Router) {}

  canActivate(next: ActivatedRouteSnapshot): UrlTree | boolean {
    return this.routeIsAvailable() || this.buildRedirectUrlTree(next);
  }

  private buildRedirectUrlTree(next: ActivatedRouteSnapshot): UrlTree {
    return this.router.createUrlTree('../default-place', {relativeTo: next});
  }

  private routeIsAvailable(): boolean {
    // check state to see if route available at this time and return true/false
    return true;
  }
}
```


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

atscott reviewed (approved):
LGTM, just one nit on a misspelled variable. Thanks for the addition!

atscott commented:
[presubmit](http://tap/402597039)

jessicajaniuk reviewed (approved):
LGTM :cookie:

This looks good aside from a typo in a variable name.

reviewed-for: public-api

daiscog committed:
feat(router): allow passing an ActivatedRouteSnapshot for relative link resolution in Router.createUrlTree

The relativeTo property of the navigationExtras parameter in the createUrlTree method of the Router class may now be an ActivatedRoute or an ActivatedRouteSnapshot.
The latter is useful in CanActivate route guards when building an alternative UrlTree to navigate to, relative to the URL that the current navigation is attempting to access.

BREAKING CHANGE: The relativeTo property of the UrlCreationOptions interface in the router package may now also be an ActivatedRouteSnapshot.

The relativeTo property is now typed `ActivatedRoute | ActivatedRouteSnapshot | null`.
Code accessing relativeTo may have to account for the possiblity this property is an ActivatedRouteSnapshot.

Fixes #22763

atscott reviewed (changes requested):
Setting my review to request changes to add test cases. From my recent investigations in this area, I unfortunately don't think this will actually achieve the intended goal.

An alternative _could_ be to expose a standalone method that creates a `UrlTree` from an `ActivatedRoute`/`ActivatedRouteSnapshot` without the `urlTree: UrlTree` parameter that's in this function. This should be possible but would require quite a bit of work and testing.

daiscog commented:
Hi @atscott 

No problem, thanks for the extra info.  I'll see if I can get time to work some more on this at the weekend.

Cheers,

Dai

atscott commented:
Closing due to https://github.com/angular/angular/pull/43800#issuecomment-954787117. I don't believe this will actually work based on how the code is structured now. Feel free to open a new PR if you have a test that proves me wrong.
